commons/c2cgeoportal_commons/alembic/static/1857owc78a07_add_last_login_and_expiration_date.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015-2018, Camptocamp SA
+# All rights reserved.
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# The views and conclusions contained in the software and documentation are those
+# of the authors and should not be interpreted as representing official policies,
+# either expressed or implied, of the FreeBSD Project.
+
+
+"""
+Add columns last_login, expire_on (both datetime) and deactivated (boolean) on
+table main_static."user".
+
+Revision ID: 1857owc78a07
+Revises: 5472fbc19f39
+Create Date: 2018-03-29 13:15:23.228907
+"""
+
+from alembic import op
+from sqlalchemy import Boolean, Column, DateTime
+from sqlalchemy.schema import MetaData, Table
+from c2cgeoportal_commons.config import config
+
+# revision identifiers, used by Alembic.
+revision = '1857owc78a07'
+down_revision = '5472fbc19f39'
+
+
+def upgrade():
+    staticschema = config['schema_static']
+
+    # Instructions
+    op.add_column('user', Column('last_login', DateTime(timezone=True)), schema=staticschema)
+    op.add_column('user', Column('expire_on', DateTime(timezone=True)), schema=staticschema)
```

Comment:
Are sou sure that we needs the timezone?

Reply:
@arnaud-morvan it is useful ? We don't need the information of timezone, no ?

Reply:
I never put a datetime in a database without timezone as I 'm not sure everybody is in the same timezone.

Reply:
but I'm not sure we need the time.

Reply:
the datetime is stored in UTC, then the time zone is ... **UTC**.
when you store a datetime with time zone, the datetime is also stored in UTC but display in local timezone.
So when you store your datetime in UTC like a datetime with time zone why don't you give the timezone ?

Reply:
Note that creating the field with timezone avoid passing a datetime in bad timezone.

Reply:
> So when you store your datetime in UTC like a datetime with time zone why don't you give the timezone ?

I don't want to store unneeded data..., and I want to be coherent in all the application...

Reply:
So, I removes it, thanks (that effectively complicate things in some cases)

Reply:
I'm sorry to insist, but this not useless data, it is important to qualify a datetime with it's timezone. This avoid doing mistakes about timezones.

Reply:
Mon voisin me dit qu'une datetime sans timezone ca ne vaut rien

Reply:
No, this is not the effective behavior with timezone enabled.
With time zone enabled, all dates are stored in UTC and returned in local timezone. This ensure compare and sort work as expected and avoid confusion about timezones.
This ensure that all datetimes are stored in UTC transparently.
So you're telling me that you do not want to set timezone to get the behavior you have with it.

Reply:
I can add that by "that effectively complicate things in some cases" i said before, I mean that it breaks tests (and probably the app) because we compare datetime with explicit timezone (the db, with timezone=True) to datetime without explicit timezone (datetime.utcnow())

If we want to use the timezone in the db, we probably need to cast all the method we use to compare time to the time that comes from db. That's probably not hard, but I'm out of time, and that was not required for this feature. And if one day we need timezone in the db, we can still do the change.

Reply:
What you say it's not what's written in Table 8-9. here: https://www.postgresql.org/docs/9.1/static/datatype-datetime.html

Reply:
=> To address these difficulties, we recommend using date/time types that contain both date and time when using time zones. 
...
All timezone-aware dates and times are stored internally in UTC. They are converted to local time in the zone specified by the timezone configuration parameter before being displayed to the client.

Reply:
If we do not set timezone in datetime fields, they will not be converted from local timezone to UTC between input and database, and not converted from UTC to local timezone for display.
We need the timezone for correct behavior between database and UI.

Reply:
And that's for 2.4...

Reply:
I just open a Jira ticket for that

Reply:
Have you tested the UI with timezone enabled ? I've done, it works perfectly with time zone enabled. Without datetimes will not be saved as UTC in the database.